### PR TITLE
23 hardware

### PR DESCRIPTION
**Pull request by TrevorEdwards**
Contains most of the hardware feature.

Currently missing backend for emails and frontend for day-of page.

### CODE REVIEW

File: models/hardware_item.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    this.quantityOwned = newQuantity;
+    this.quantityAvailable = this.quantityOwned - itemsOut;
+
+    if (this.quantityOwned < this.quantityAvailable) {
+        return callback('Quantity available exceeds quantity owned!');
+    }
+    this.save(callback);
+};
+
+hardwareItemSchema.methods.transaction = function transaction(changeInQuantity, callback) {
```
Comment (mahsu) [inline models/hardware_item.js]: Name with a verb: `changeQuantity`?

File: models/hardware_item.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    } else if (this.quantityAvailable < 0) {
+        return callback('Negative quantity available!');
+    }
+
+    this.save(callback);
+};
+
+// Hooks
+hardwareItemSchema.pre('save', function(next) {
+    if (this.quantityAvailable > this.quantityOwned) {
```
Comment (mahsu) [inline models/hardware_item.js]: Looks like this is reused in different places - maybe refactor to `checkQuantityInvariant` or something

File: routes/admin.js [hunk 3]
```diff
@@ -611,6 +614,30 @@ router.get('/stats', function (req, res, next) {
...
+    router.get('/hardware', function (req, res, next) {
+        async.parallel({
+            inventory: function inventory(cb) {
+                HardwareItem.find({}, null, {sort: {name: 'asc'}}, cb);
+            },
+            checkouts: function checkouts(cb) {
+                HardwareItemCheckout.find().populate('student_id inventory_id').exec(cb);
+            },
+            transactions: function transactions(cb) {
+                HardwareItemTransaction.find().populate('studentId').exec(cb);
```
Comment (mahsu) [inline routes/admin.js]: is there a difference between `studentId` and `student_id` in 628? Might want to keep naming consistent if there isn't

File: routes/api/admin.js [hunk 4]
```diff
@@ -916,6 +921,174 @@ function rsvpDeadlineOverride(req, res, next) {
...
+        } else if (!result.item) {
+            return res.status(500).send('No such item');
+        }
+
+        InventoryTransaction.findOne({student: result.student.id, inventory_id: result.item.id}, function (err, transaction) {
+            if (err) {
+                return res.status(500).send(err);
+            }
+
+            if (body.checkingOut) {
```
Comment (TrevorEdwards) [inline routes/api/admin.js]: Do you mean this for transactions?

Comment (mahsu) [inline routes/api/admin.js]: Yeah. I guess it's fine for now

File: routes/api/admin.js [hunk 5]
```diff
@@ -916,6 +921,174 @@ function rsvpDeadlineOverride(req, res, next) {
...
+                    return res.status(500).send('Quantity exceeds availability');
+                }
+            } else {
+                if (!transaction) {
+                    return res.status(500).send('User has no transactions for that item.');
+                } else if (body.quantity > transaction.quantity) {
+                    return res.status(500).send('User has not checked out that many items of that type!');
+                }
+
+                if (result.item.quantityAvailable + body.quantity >= result.item.quantityOwned) {
```
Comment (TrevorEdwards) [inline routes/api/admin.js]: Fixed in presave check

**Comment by TrevorEdwards** [comment]
Going to focus on testing this now

**Comment by TrevorEdwards** [comment]
Actually, formal tests are a bit hard to write for this since it requires a standard user across all environments. We should probably just test this in rc for now

File: routes/api/admin.js [hunk 6]
```diff
@@ -947,6 +952,207 @@ function rsvpDeadlineOverride(req, res, next) {
...
+ * @apiParam {String} email Email of the student for the transaction
+ * @apiParam {Number} quantity The quantity of hardware to transact
+ * @apiParam {String} name The unique name of the hardware being transacted
+ **/
+function transactHardware({body}, res, next) {
+    if (body.checkingOut === undefined || !body.email || body.quantity === undefined || !body.name) {
+        return res.status(500).send('Missing a parameter, check the API!');
+    }
+
+    body.quantity = Number(body.quantity); // This formats as a string by default
```
Comment (mahsu) [inline routes/api/admin.js]: I think you need to catch the NaN case.

File: models/hardware_item.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    this.quantityOwned = newQuantity;
+    this.quantityAvailable = this.quantityOwned - itemsOut;
+
+    if (this.quantityOwned < this.quantityAvailable) {
+        return callback('Quantity available exceeds quantity owned!');
+    }
+    this.save(callback);
+};
+
+hardwareItemSchema.methods.changeQuantity = function changeQuantity(changeInQuantity, callback) {
```
Comment (mahsu) [inline models/hardware_item.js]: I would rename to addQuantity since it's a relative change

File: models/hardware_item.js [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        return callback('Negative quantity available!');
+    }
+
+    this.save(callback);
+};
+
+// Hooks
+hardwareItemSchema.pre('save', function checkQuantityInvariant(next) {
+    if (this.quantityAvailable > this.quantityOwned) {
+        return next(new Error('Quantity Available Exceeds Quantity Owned'));
```
Comment (mahsu) [inline models/hardware_item.js]: Will using `Error` here crash the app?

File: models/hardware_item_transaction.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
+"use strict";
+/**
+ * Holds inventory information for a hardware item
```
Comment (mahsu) [inline models/hardware_item_transaction.js]: Update comment

File: models/hardware_item_checkout.js [hunk 10]
```diff
@@ -0,0 +1,23 @@
+"use strict";
```
Comment (mahsu) [inline models/hardware_item_checkout.js]: This seems redundant with `hardware_item_transaction`

File: models/hardware_item_checkout.js [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+/**
+ * Tracks a checkout of hardware until it is checked back in
+ */
+
+var mongoose = require('mongoose');
+
+var hardwareItemCheckoutSchema = new mongoose.Schema({
+    student_id: {type: mongoose.Schema.Types.ObjectId, ref: "User", required: true},
+    inventory_id: {type: mongoose.Schema.Types.ObjectId, ref: "HardwareItem", required: true},
+    quantity: {type: Number, required: true}
```
Comment (mahsu) [inline models/hardware_item_checkout.js]: do we want to add a timestamp here?

File: routes/api/admin.js [hunk 12]
```diff
@@ -947,6 +952,207 @@ function rsvpDeadlineOverride(req, res, next) {
...
+ * @apiParam {Number} quantity The quantity of hardware we own
+ * @apiParam {String} name The unique name of the hardware
+ **/
+function setInventory(req, res, next) {
+    let body = req.body;
+    if (!body || !body.quantity || !body.name) {
+        return res.status(500).send('Missing quantity or name');
+    }
+
+    if (body.quantity <= 0) {
```
Comment (mahsu) [inline routes/api/admin.js]: Do you need to cast body.quantity to a number?

File: routes/api/admin.js [hunk 13]
```diff
@@ -947,6 +952,207 @@ function rsvpDeadlineOverride(req, res, next) {
...
+            return res.status(500).send(err);
+        } else if (!result.student) {
+            return res.status(500).send('No such user');
+        } else if (!result.item) {
+            return res.status(500).send('No such item');
+        }
+
+        InventoryTransaction.findOne({
+            student_id: result.student.id,
+            inventory_id: result.item.id
```
Comment (mahsu) [inline routes/api/admin.js]: are these supposed to be `._id`? or is `.id` an alias?

File: routes/api/admin.js [hunk 14]
```diff
@@ -947,6 +952,207 @@ function rsvpDeadlineOverride(req, res, next) {
...
+                                        return res.redirect('/admin/hardware');
+                                    });
+                                });
+                            }
+                    });
+                });
+            }
+        });
+    });
+}
```
Comment (mahsu) [inline routes/api/admin.js]: i think we should cleanup this method in the future. There's a decent bit of model logic.

File: routes/index.js [hunk 15]
```diff
@@ -95,4 +99,15 @@ router.get('/wayel', function (req, res, next) {
     });
 });
 
+/**
+ * @api {GET} /hardware hardware chcekout page
+ * @apiName Hardware
+ * @apiGroup Index
+ */
+router.get('/hardware', function (req, res, next) {
```
Comment (mahsu) [inline routes/index.js]: Shouldn't this be in admin? The hardware lab will always be staffed by a member of the organizing team.

**Comment by mahsu** [comment]
Looks like there's merge conflicts.

File: models/hardware_item.js (hunk 8 above)
Comment (TrevorEdwards) [inline models/hardware_item.js]: I don't believe so.

http://mongoosejs.com/docs/middleware.html shows that it is idiomatic to use next(new Error) in these cases, and it's just handled with an if (err) console.error.. thing later

File: models/hardware_item_checkout.js (hunk 10 above)
Comment (TrevorEdwards) [inline models/hardware_item_checkout.js]: They are different. Maybe my comment update will help differentiate them.

File: models/hardware_item_checkout.js (hunk 11 above)
Comment (TrevorEdwards) [inline models/hardware_item_checkout.js]: No- The associated transaction is sufficient to track the timestamp

File: routes/api/admin.js (hunk 12 above)
Comment (TrevorEdwards) [inline routes/api/admin.js]: I will just in case!

File: routes/api/admin.js (hunk 13 above)
Comment (TrevorEdwards) [inline routes/api/admin.js]: id is a virtual getter for _id for any mongoose model

**Comment by mahsu** [comment]
lgtm

**Pull request merged**